### PR TITLE
Fix and update cotire

### PR DESCRIPTION
**Pull request by gassmoeller**
This is a workaround for #2314. It updates cotire to the latest release, and disables it for cmake version >= 3.9 due to the errors with the generator expressions that are used in deal.II as discussed in that issue. If we fix the issues otherwise we can again enable cotire for later cmake versions.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -393,15 +393,19 @@ ENDIF()
 
 # Check if we want to use cotire to precompile header files and prepare a
 # unity build target. This speeds up compile time, but can fail on some
-# machines, therefore it is disabled by default.
-SET(ASPECT_PRECOMPILE_HEADERS OFF CACHE BOOL "Precompile external header files using cotire to speedup compile time.")
```
Comment (tjhei) [inline CMakeLists.txt]: I think you should always define this variable and force it to off, like this:
```
SET(ASPECT_PRECOMPILE_HEADERS OFF CACHE BOOL "Precompile external header files using cotire to speedup compile time.")
IF(CMAKE_VERSION VERSION_LESS 3.9)
... existing code
ELSE()
  SET(ASPECT_PRECOMPILE_HEADERS OFF CACHE BOOL "" FORCE)
ENDIF()
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -486,7 +490,7 @@ IF (ASPECT_USE_FP_EXCEPTIONS)
   ENDIF()
 ENDIF()
 
-IF (ASPECT_PRECOMPILE_HEADERS)
+IF (CMAKE_VERSION VERSION_LESS 3.9 AND ASPECT_PRECOMPILE_HEADERS)
```
Comment (tjhei) [inline CMakeLists.txt]: then you don't need this change

**Comment by tjhei** [comment]
This should go into 2.1...

**Comment by gassmoeller** [comment]
Like this? I added it to the description of the parameter to avoid confusion why this variable keeps switching itself off.

**Comment by tjhei** [comment]
> I added it to the description of the parameter to avoid confusion why this variable keeps switching itself off.

Yes, this is what I thought. You make a good point: we could also report an error if somebody enables cotire and has a cmake version that is too new. 
```
SET(ASPECT_PRECOMPILE_HEADERS OFF CACHE BOOL "Precompile external header files using cotire to speedup compile time. Currently only supported for cmake versions older than 3.9.")
￼	IF(CMAKE_VERSION VERSION_LESS 3.9)
￼	  IF (ASPECT_PRECOMPILE_HEADERS)
￼	    # For the unity build we need to compile helper_functions.cc before core.cc
￼	    # ensure this by reordering the files
￼	    FIND_FILE(HELPER_PATH helper_functions.cc HINTS "${CMAKE_SOURCE_DIR}/source/simulator")
￼	    LIST(REMOVE_ITEM TARGET_SRC ${HELPER_PATH})
￼	    LIST(INSERT TARGET_SRC 0 ${HELPER_PATH})
￼	  ENDIF()
￼	ELSE()
￼	  MESSAGE(FATAL "Sorry, you need CMake version 3.9 or less for this feature to work.")
￼	ENDIF()
```
I leave this to you.

**Comment by tjhei** [comment]
Looks good. Ready to merge when the testers are happy.

**Pull request merged**